platform/android/src/android_renderer_frontend.cpp

```diff
@@ -1,72 +1,132 @@
 #include "android_renderer_frontend.hpp"
 
-#include <mbgl/renderer/backend_scope.hpp>
-#include <mbgl/renderer/renderer.hpp>
+#include <mbgl/actor/scheduler.hpp>
+#include <mbgl/renderer/renderer_observer.hpp>
+#include <mbgl/storage/file_source.hpp>
+#include <mbgl/util/thread.hpp>
+#include <mbgl/util/run_loop.hpp>
+#include <mbgl/util/logging.hpp>
+
+#include "android_renderer_backend.hpp"
 
 namespace mbgl {
 namespace android {
 
-AndroidRendererFrontend::AndroidRendererFrontend(
-        std::unique_ptr<Renderer> renderer_,
-        RendererBackend& backend_,
-        InvalidateCallback invalidate)
-        : renderer(std::move(renderer_))
-        , backend(backend_)
-        , asyncInvalidate([=, invalidate=std::move(invalidate)]() {
-            invalidate();
-        }) {
+// Forwards RendererObserver signals to the given
+// Delegate RendererObserver on the given RunLoop
+class ForwardingRendererObserver : public RendererObserver {
+public:
+    ForwardingRendererObserver(util::RunLoop& mapRunLoop_, RendererObserver& delegate_)
+            : mapRunLoop(mapRunLoop_)
+            , delegate(delegate_) {}
+
+    void onInvalidate() override {
+        mapRunLoop.invoke([&]() {
+            delegate.onInvalidate();
+        });
+    }
+
+    void onResourceError(std::exception_ptr err) override {
+        mapRunLoop.invoke([&, err]() {
+            delegate.onResourceError(err);
+        });
+    }
+
+    void onWillStartRenderingMap() override {
+        mapRunLoop.invoke([&]() {
+            delegate.onWillStartRenderingMap();
+        });
+    }
+
+    void onWillStartRenderingFrame() override {
+        mapRunLoop.invoke([&]() {
+            delegate.onWillStartRenderingFrame();
+        });
+    }
+
+    void onDidFinishRenderingFrame(RenderMode mode, bool repaintNeeded) override {
+        mapRunLoop.invoke([&, mode, repaintNeeded]() {
+            delegate.onDidFinishRenderingFrame(mode, repaintNeeded);
+        });
+    }
+
+    void onDidFinishRenderingMap() override {
+        mapRunLoop.invoke([&]() {
+            delegate.onDidFinishRenderingMap();
+        });
+    }
+
+private:
+    util::RunLoop& mapRunLoop;
+    RendererObserver& delegate;
+};
+
+AndroidRendererFrontend::AndroidRendererFrontend(MapRenderer& mapRenderer_)
+        : mapRenderer(mapRenderer_)
+        , asyncInvalidate([&]() {
```

Comment:
Do we need the async? I would expect `GLSurfaceView.requestRender` to coalesce internally.

Reply:
No, you're right. Removed this. I left it in initially to skip some calls over the JNI boundary.